### PR TITLE
docs: add coding style guide for contributors

### PR DESCRIPTION
**Pull request by incandescentxxc**
## Motivation

Give human contributors and coding agents a shared, repository-local coding style guide while keeping agent instructions short and separate from the external documentation wiki.

## Modifications

- Add `contributing/coding-style.md` covering simplicity, naming, typing, structure, error handling, logging, configuration, imports, training, comments, and pre-commit checks.
- Reference the guide from `AGENTS.md` and point `CLAUDE.md` to `AGENTS.md`.

## Related Issues

Adapted from the coding-style prompt and anti-pattern examples in https://github.com/zhaochenyang20/sglang-diffusion-routing/issues/32. This PR does not close that issue.

## Accuracy Test

Documentation only; model accuracy tests and unit tests are not applicable.

Validation:
- `git diff --cached --check` passed before commit.
- Checked that all local Markdown links in the three new files resolve.
- `pre-commit run --all-files` passed with pre-commit 4.6.2; all applicable hooks passed, including Rust formatting.

## Benchmark & Profiling

Not applicable; no runtime behavior changes.

## Checklist

- [x] Format your code according with pre-commit.
- [ ] Add unit tests. (Not applicable: documentation only.)
- [x] Update documentation / docstrings / example tutorials as needed.
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed. (Not applicable.)

### CODE REVIEW

File: .claude/skills/code-review/coding-style.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+### FILE STRUCTURE
+
+- File > ~400 lines → extract a module. But a 30-line file holding one `_helper` called once
+  is also wrong — merge it. Related functions belong together; one concern per file
+  does not mean one function per file.
+- One entry-point mechanism per repo (hydra/argparse/fire/`[project.scripts]`). `if __name__
+  == "__main__"` only in entry-point modules, never library modules.
+- Module-level side effects (env mutation, global state, `setup_root`, resolver registration)
+  only in entry-point modules. Library imports must be side-effect-free.
+- No dead code: unreachable branches, commented-out blocks, stale TODOs,
```
Comment (ychu824) [inline .claude/skills/code-review/coding-style.md:103]: "No dead code", "unreachable branches": Does it contradict with "Every `if` must have an `else`"?

File: .claude/skills/code-review/coding-style.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+- Either `requires-python >= 3.10` (native `X | Y`) or `from __future__ import annotations`.
+  Don't use `Optional` to work around forward refs — use quoted annotations
+  (`"ModelConfig"`). Type-only imports are covered in IMPORTS.
+- Closed value sets → `Literal[...]` or `Enum`, not bare strings in comparisons.
+- No mutable function defaults: `def f(x=[])`/`= {}` are bugs. Use a `None` sentinel.
+- Use concrete types, including model and decoder types, rather than `any`/`Any`
+  or bare `dict`/`list`/`tuple`. Annotate structured values and resource handles
+  according to their actual contracts, including element types and optionality.
+- Do not accept a parameter only to immediately delete it to silence type or lint
+  checks, such as starting a function with `del request_id`. Remove unnecessary
```
Comment (ychu824) [inline .claude/skills/code-review/coding-style.md:81]: Maybe include the example signature is better:
``` python
def xxx(...,
    request_id:str,
    ...
):
    del request_id
```

**Review by taylorty — COMMENTED** [review]
Useful guide overall, and several rules do match this repo (stdlib `logging`, `X | None` typing, the pre-commit link). A few things should be fixed before merge.

**Blocking — `contributing/coding-style.md:128`, "Every `if` must have an `else`."** 7,491 of 8,522 `if` statements under `sglang_omni/` (88%) have no `else`/`elif`, and the rule contradicts three others in this same document: "Match the surrounding codebase's conventions" (:8), "Keep related checks flat and minimize nesting" (:130 — guard clauses take no `else`), and "No dead code: unreachable branches" (:99). Followed literally via `AGENTS.md`, agents will staple `else: pass` onto every guard clause. Suggest narrowing to: *`if`/`elif` chains that select a value or dispatch on a closed set must be exhaustive; guard clauses and early returns take no `else`.*

**Tooling that doesn't exist here.** `:164` recommends `pyrootutils.setup_root(..., indicator=".project-root", ...)` — not a dependency, not imported anywhere, and there is no `.project-root` file. Same shape at `:93` (entry points — this repo uses `[project.scripts]` → `sgl-omni = "sglang_omni.cli:app"`), `:134` (loguru isn't a dep; the repo is stdlib `logging`), and `:148` (config mechanism). Since this is the repo-local guide, please name this repo's actual choice instead of listing options.

**`:180-183` TRAINING** — there is no training code in sglang-omni; it's an inference serving framework. A prescribed `train`/`trainer` split is the speculative generality that `:14` bans.

**`:18` / `:23` / `:29` — three thresholds for one decision.** Extract a helper at "at least twice" (≥2), don't pre-extract unless "more than twice" (>2), inline when "once or twice" (≤2). At exactly two call sites the doc says extract, don't extract, and inline. Pick one number.

Minor:
- `:71` — `any` is the builtin function; you mean `Any`. "model and decoder types" is vocabulary from the source repo that maps onto nothing here.
- `:90` (~400-line files) and `:124` (no `getattr` defaults) are contradicted by 128 files and 607 call sites respectively. Fine as forward-looking rules — add a line scoping the guide to new and modified code.
- Placement: top-level `contributing/` is outside the Sphinx tree, unlinked from `README.md` and the PR template, so human contributors won't find it. Consider `docs/developer_reference/` alongside the other contributor docs, or add links.
- Headings jump `#` → `###` with no `##`.

File: .claude/skills/code-review/coding-style.md (hunk 2 above)
Comment (incandescentxxc) [reply .claude/skills/code-review/coding-style.md:81]: my two cents are keeping this coding-style succient and not carrying too many verbose examples. Let's see how it goes and we can revisit that.

File: contributing/coding-style.md [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+- ONE logging library per repo. Don't mix stdlib `logging` and `loguru`. Configure once.
+- One logger per module: `logging.getLogger(__name__)` or `from loguru import logger`.
+- Keep messages terse. No manual `[Info]`, `[Warn]`, or `[step N]` prefixes;
+  use the logger's level and context fields.
+- `print` only for CLI output the end user reads (`--help`, visualization, `__main__` demo).
+  Runtime info — even debug — goes through the logger.
+- Distributed: guard rank-zero via logger config or a `RankedLogger`-style adapter, not
+  scattered `if is_main:`. (Training loops keep theirs; don't spread the pattern.)
+- f-string vs `%-style`: pick one per repo. loguru → f-strings idiomatic; stdlib hot path →
```
Comment (suninnn) [inline contributing/coding-style.md]: Would it be more accurate to recommend Loguru’s native {} formatting rather than f-strings?

File: .claude/skills/code-review/coding-style.md [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+- Define a constant used by only one module in that module; do not create a
+  cross-file import solely for it.
+
+### IMPORTS
+
+- Group: stdlib / third-party / local, blank-line separated, alphabetical within group.
+- No `sys.path.insert` hacks. Use package imports, PYTHONPATH, or
+  `pyrootutils.setup_root(__file__, indicator=".project-root", pythonpath=True)`
+  in entry-point modules only. Use one setup mechanism per repo.
+- No function-local imports except documented circular-dependency breaks. Put
```
Comment (suninnn) [inline .claude/skills/code-review/coding-style.md]: #32 allows lazy imports for heavy optional dependencies. Should we allow that exception here as well?

File: .claude/skills/code-review/coding-style.md (hunk 1 above)
Comment (incandescentxxc) [reply .claude/skills/code-review/coding-style.md:103]: i think the uncreachable branches referst to the scenario where we over-defensively have an "else" or "catch" block.

File: .claude/skills/code-review/coding-style.md [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+  type-only cycle-breaking imports under `if TYPE_CHECKING:` with quoted annotations.
```
Comment (Zhufeng-Qiu) [inline .claude/skills/code-review/coding-style.md]: Location: `contributing/coding-style.md:166–167` — IMPORTS

Could we allow function-local imports for optional dependencies? [Audar-TTS](https://github.com/sgl-project/sglang-omni/blob/42f16e9aebffa8a50c4b9f13cc91f49f373ba875/sglang_omni/models/audar_tts/stages.py#L51-L58) already uses this to defer loading NeuCodec.

Suggested wording: “Prefer module-level imports; allow function-local imports for optional dependencies, necessary initialization ordering, or documented circular-dependency breaks.”

File: .claude/skills/code-review/coding-style.md [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+- English only: comments, docstrings, log strings, CLI help. (User-facing translatable
+  strings go through i18n — not this rule's concern.) If the repo is non-English, match it.
+- Comments sparse, one-line, only for the genuinely non-obvious. Restating code is noise.
+- NO process markers: no ★, `# P1`, `# [FIX]`, `# TODO` without a ticket, `# === SECTION ===`
+  banners.
+- NO provenance leakage: never name other repos/upstream/"the closed source" in source.
+- NO verbose rationale in comments — the "why" lives in design docs / commits / PRs.
+- Docstrings: Google-style, 1-3 lines. Args/Returns only when non-obvious. One short module
+  docstring per file.
+- Do not add opening comment blocks or lengthy explanations after class definitions.
```
Comment (Zhufeng-Qiu) [inline .claude/skills/code-review/coding-style.md:49]: Location: `contributing/coding-style.md:39–43` — LANGUAGE & COMMENTS

Could we exempt license/copyright notices and upstream attribution from these restrictions? The [CosyVoice3 MLX header](https://github.com/sgl-project/sglang-omni/blob/42f16e9aebffa8a50c4b9f13cc91f49f373ba875/sglang_omni/models/fun_cosyvoice3/mlx/vocoder/flow_matching.py#L1-L5) records its source repository and commit for traceability.

Suggested modification: """
- Avoid irrelevant provenance commentary; preserve license/copyright notices and
  concise upstream attribution.
- NO verbose rationale in comments — the "why" lives in design docs / commits / PRs.
- Docstrings: Google-style, 1-3 lines. Args/Returns only when non-obvious. One short module
  docstring per file.
- Do not add opening comment blocks, except for license/copyright notices and
  concise upstream attribution. Avoid lengthy explanations after class definitions.
"""

**Comment by incandescentxxc** [comment]
> Useful guide overall, and several rules do match this repo (stdlib `logging`, `X | None` typing, the pre-commit link). A few things should be fixed before merge.
> 
> **Blocking — `contributing/coding-style.md:128`, "Every `if` must have an `else`."** 7,491 of 8,522 `if` statements under `sglang_omni/` (88%) have no `else`/`elif`, and the rule contradicts three others in this same document: "Match the surrounding codebase's conventions" (:8), "Keep related checks flat and minimize nesting" (:130 — guard clauses take no `else`), and "No dead code: unreachable branches" (:99). Followed literally via `AGENTS.md`, agents will staple `else: pass` onto every guard clause. Suggest narrowing to: _`if`/`elif` chains that select a value or dispatch on a closed set must be exhaustive; guard clauses and early returns take no `else`._
> 
> **Tooling that doesn't exist here.** `:164` recommends `pyrootutils.setup_root(..., indicator=".project-root", ...)` — not a dependency, not imported anywhere, and there is no `.project-root` file. Same shape at `:93` (entry points — this repo uses `[project.scripts]` → `sgl-omni = "sglang_omni.cli:app"`), `:134` (loguru isn't a dep; the repo is stdlib `logging`), and `:148` (config mechanism). Since this is the repo-local guide, please name this repo's actual choice instead of listing options.
> 
> **`:180-183` TRAINING** — there is no training code in sglang-omni; it's an inference serving framework. A prescribed `train`/`trainer` split is the speculative generality that `:14` bans.
> 
> **`:18` / `:23` / `:29` — three thresholds for one decision.** Extract a helper at "at least twice" (≥2), don't pre-extract unless "more than twice" (>2), inline when "once or twice" (≤2). At exactly two call sites the doc says extract, don't extract, and inline. Pick one number.
> 
> Minor:
> 
> * `:71` — `any` is the builtin function; you mean `Any`. "model and decoder types" is vocabulary from the source repo that maps onto nothing here.
> * `:90` (~400-line files) and `:124` (no `getattr` defaults) are contradicted by 128 files and 607 call sites respectively. Fine as forward-looking rules — add a line scoping the guide to new and modified code.
> * Placement: top-level `contributing/` is outside the Sphinx tree, unlinked from `README.md` and the PR template, so human contributors won't find it. Consider `docs/developer_reference/` alongside the other contributor docs, or add links.
> * Headings jump `#` → `###` with no `##`.


About `if` must have an `else`, I think a better representaton which can generalize the intention is the following 

```
Validate inputs and preconditions before the main logic. When possible, organizing
 conditions into clear, mutually exclusive if/elif/else branches. Return early for 
 invalid cases, and keep the main execution path in the final branch to avoid unnecessary 
 lookups, repeated checks, and deeply nested logic.
```

I fixed the rest

File: contributing/coding-style.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+- Group: stdlib / third-party / local, blank-line separated, alphabetical within group.
+- No `sys.path.insert` hacks. Use package imports, PYTHONPATH, or
+  `pyrootutils.setup_root(__file__, indicator=".project-root", pythonpath=True)`
+  in entry-point modules only. Use one setup mechanism per repo.
+- No function-local imports except documented circular-dependency breaks. Put
+  type-only cycle-breaking imports under `if TYPE_CHECKING:` with quoted annotations.
+- Import from the defining module using the full package path, such as
+  `from xxx.yy.zzz import kkk`, rather than through `__init__.py`. Keep package
+  re-exports minimal and define an explicit `__all__`; no wildcard imports.
```
Comment (Zhufeng-Qiu) [inline contributing/coding-style.md]: Location: `contributing/coding-style.md:168–170` — IMPORTS

Minor: could we scope the defining-module import rule to repository-internal imports? We already use package-level public APIs such as [`from pydantic import BaseModel`](https://github.com/sgl-project/sglang-omni/blob/42f16e9aebffa8a50c4b9f13cc91f49f373ba875/sglang_omni/serve/protocol.py#L11).

Suggested modification: "For repository-internal imports, use the defining module. For third-party libraries, prefer their documented public import paths."

File: contributing/coding-style.md (hunk 3 above)
Comment (incandescentxxc) [reply contributing/coding-style.md]: this repo is using stdlib logging, i excluded the Logruru related recommendation

File: .claude/skills/code-review/coding-style.md (hunk 4 above)
Comment (incandescentxxc) [reply .claude/skills/code-review/coding-style.md]: sgtm, fixed

File: .claude/skills/code-review/coding-style.md (hunk 5 above)
Comment (incandescentxxc) [reply .claude/skills/code-review/coding-style.md]: sgtm, fixed

**Review by Zhufeng-Qiu — COMMENTED** [review]
Thanks for the update! I rechecked `199e9dc`; both import-related concerns are addressed. 

One small follow-ups in `.claude/skills/code-review/coding-style.md`:
- Line 181: the pre-commit link should be `../../../.pre-commit-config.yaml`; the current relative path points to a nonexistent file.

**Review by zhaochenyang20 — APPROVED** [review]

File: .claude/skills/code-review/coding-style.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+- Do not turn execution failures into fabricated successful outputs. When a
+  function guarantees a return type, use its result directly instead of adding
+  impossible `None` checks or redundant type coercion. Validate genuinely untrusted
+  results at their boundary instead of weakening an established internal contract.
+- Access fields on known types directly. Do not use `getattr` defaults or `hasattr`
+  probes to hide missing required attributes, or catch attribute errors for
+  operations guaranteed by the interface. Missing required attributes should fail
+  visibly rather than trigger a search for alternate fields or fallback defaults.
+
+## CONTROL FLOW
```
Comment (Hoiyeuhng) [inline .claude/skills/code-review/coding-style.md:132]: **suggestion (non-blocking)**: Could we add this under **CONTROL FLOW**?

“Assign a value to a variable before reading it. If the assignment is inside an `if` or `try` block, make sure it runs before the variable is used.”

For example, `TorchProfiler.start()` reads `rank` at [line 58](https://github.com/sgl-project/sglang-omni/blob/b29084bfa5b80779ea8387f3da951b5b0c506c8c/sglang_omni/profiler/torch_profiler.py#L58) or [line 62](https://github.com/sgl-project/sglang-omni/blob/b29084bfa5b80779ea8387f3da951b5b0c506c8c/sglang_omni/profiler/torch_profiler.py#L62), but assigns its value at [line 76](https://github.com/sgl-project/sglang-omni/blob/b29084bfa5b80779ea8387f3da951b5b0c506c8c/sglang_omni/profiler/torch_profiler.py#L76). When a profiler is already running, this causes `UnboundLocalError`.

Comment (incandescentxxc) [reply .claude/skills/code-review/coding-style.md:132]: oh to me this sounds more like a correctness issue than a coding style? i wonder why unit tests haven't got this. wdyt?

Comment (Hoiyeuhng) [reply .claude/skills/code-review/coding-style.md:132]: yeah....  You're right, I overthought this. I found it while enabling Ruff's default checks, which weren't enabled for this repo in our existing hook. This belongs in lint checks and regression tests rather than the coding style guide. Happy to leave it out here.

**Comment by incandescentxxc** [comment]
> Thanks for the update! I rechecked `199e9dc`; both import-related concerns are addressed.
> 
> One small follow-ups in `.claude/skills/code-review/coding-style.md`:
> 
> * Line 181: the pre-commit link should be `../../../.pre-commit-config.yaml`; the current relative path points to a nonexistent file.

addressed this using a project-root based path ([reference](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax): Links starting with / will be relative to the repository root.)
